src/linear_algebra/char_poly.lean

```diff
@@ -0,0 +1,107 @@
+-/
+
+noncomputable theory
+
+universes u v w
+
+open polynomial matrix
+open_locale big_operators
+
+variables {R : Type u} [comm_ring R]
+variables {n : Type w} [fintype n] [decidable_eq n]
+
+open finset
+
+/--
+The "characteristic matrix" of `M : matrix n n R` is the matrix of polynomials $t I - M$.
+The determinant of this matrix is the characteristic polynomial.
+-/
+def char_matrix (M : matrix n n R) : matrix n n (polynomial R) :=
+matrix.scalar n (X : polynomial R) - (C : R →+* polynomial R).map_matrix M
+
+@[simp] lemma char_matrix_apply_eq (M : matrix n n R) (i : n) :
+  char_matrix M i i = (X : polynomial R) - C (M i i) :=
+by simp only [char_matrix, sub_left_inj, pi.sub_apply, scalar_apply_eq,
+  ring_hom.map_matrix_apply, map_apply]
+
+@[simp] lemma char_matrix_apply_ne (M : matrix n n R) (i j : n) (h : i ≠ j) :
+  char_matrix M i j = - C (M i j) :=
+by simp only [char_matrix, pi.sub_apply, scalar_apply_ne _ _ _ h, zero_sub,
+  ring_hom.map_matrix_apply, map_apply]
+
+lemma mat_poly_equiv_char_matrix (M : matrix n n R) :
+  mat_poly_equiv (char_matrix M) = X - C M :=
+begin
+  ext k i j,
+  simp only [mat_poly_equiv_coeff_apply, coeff_sub, pi.sub_apply],
+  by_cases h : i = j,
+  { subst h, rw [char_matrix_apply_eq, coeff_sub],
+    simp only [coeff_X, coeff_C],
+    split_ifs; simp, },
+  { rw [char_matrix_apply_ne _ _ _ h, coeff_X, coeff_neg, coeff_C, coeff_C],
+    split_ifs; simp [h], }
+end
+
+/--
+The characteristic polynomial of a matrix `M` is given by $\det (t I - M)$.
+-/
+def char_poly (M : matrix n n R) : polynomial R :=
+(char_matrix M).det
+
+/--
+The Cayley-Hamilton theorem, that the characteristic polynomial of a matrix,
+applied to the matrix itself, is zero.
+
+This holds over any commutative ring.
+-/
+-- This proof follows http://drorbn.net/AcademicPensieve/2015-12/CayleyHamilton.pdf
+theorem char_poly_map_eval_self (M : matrix n n R) :
+  ((char_poly M).map (algebra_map R (matrix n n R))).eval M = 0 :=
+begin
+  -- We begin with the fact $χ_M(t) I = adjugate (t I - M) * (t I - M)$,
+  -- as an identity in `matrix n n (polynomial R)`.
+  have h : (char_poly M) • (1 : matrix n n (polynomial R)) =
+    adjugate (char_matrix M) * (char_matrix M) :=
+    (adjugate_mul _).symm,
+  -- Using the algebra isomorphism `matrix n n (polynomial R) ≃ₐ[R] polynomial (matrix n n R)`,
+  -- we have the same identity in `polynomial (matrix n n R)`.
+  apply_fun mat_poly_equiv at h,
+  simp only [mat_poly_equiv.map_mul,
+    mat_poly_equiv_char_matrix] at h,
+  -- Because the coefficient ring `matrix n n R` is non-commutative,
+  -- evaluation at `M` is not multiplicative.
+  -- However, any polynomial which is a product of the form $N * (t I - M)$
+  -- is sent to zero, because the evaluation function puts the polynomial variable
+  -- to the right of any coefficients, so everything telescopes.
+  apply_fun (λ p, p.eval M) at h,
+  rw eval_mul_X_sub_C at h,
```

Comment:
Is there any reason not to write `rwa [eval_mul_X_sub_C,  mat_poly_equiv_smul_one] at h,` instead of those three lines? Is that because of the narrative in the comments?

Reply:
I think that this proof was indeed un-golfed for readability reasons.

Reply:
I wonder whether we should care about the fact that `mathcomp` people always point out how [their proof](https://github.com/math-comp/math-comp/blob/3728862662bd0a5b836dfa746921954604d051ec/mathcomp/algebra/mxpoly.v#L416-L422) is only three lines long.

Maybe we can add a comment below our proof saying:
```
theorem char_poly_map_eval_self (M : matrix n n R) :
  ((char_poly M).map (algebra_map R (matrix n n R))).eval M = 0 :=
begin
  have h : (mat_poly_equiv ((char_poly M) • 1)).eval M = (mat_poly_equiv  (adjugate (char_matrix M) * (char_matrix M))).eval M := congr_arg _ (congr_arg _ (adjugate_mul _).symm),
  simpa only [mat_poly_equiv.map_mul, mat_poly_equiv_char_matrix, eval_mul_X_sub_C, mat_poly_equiv_smul_one] using h,
end
```

Reply:
How about `by simpa ... using ...`?
Sounds good to me...

Reply:
Indeed, on Scott's screen you could use 
```
by simpa only [mat_poly_equiv.map_mul, mat_poly_equiv_char_matrix, eval_mul_X_sub_C, mat_poly_equiv_smul_one] using (congr_arg _ (congr_arg _ (adjugate_mul _).symm) : (mat_poly_equiv ((char_poly M) • 1)).eval M = (mat_poly_equiv  (adjugate (char_matrix M) * (char_matrix M))).eval M)
```